### DISCUSSION TITLE
0.14 想法整理

### DISCUSSION THREAD

**Discussion by syimyuzya** [nk2028/qieyun-js#34] [Ideas]
目前想到的方面：

- `推導方案` API 分離為獨立 package，從 `qieyun` 中移出
- 資料查詢部分重構
- (2022-06-14 增加、2022-07-31、2022-08-12 再增加）音韻地位若干調整
- 不再完整支持表達其他家分析體系，其他體系音韻地位可享受的功能將大幅受限

### `Qieyun.推導方案` 移出

原因是 qieyun-autoderiver 0.2 引入了更強（細節論起來也更複雜）的功能，該部分打算移出去獨立維護。

新的 package 大概會叫 `qieyun-deriver-tools` 之類的，這個庫的作用在於統一 `qieyun-examples`（從 JS 調用）與 `qieyun-autoderiver`（在 GUI 使用）兩者的共通部分，以及為了讓 `qieyun-examples` 的 JS API 更好用。

### 資料

這部分是為了從「資料條目 = 廣韻條目」的束縛中脫離（畢竟項目叫 `qieyun` 嘛233（僅管相當一段時間裡可能仍然以廣韻資料為絕對主體233

具體來說，原本用於《廣韻》的資料查詢功能挪至 `Qieyun.資料.廣韻`，而原 `Qieyun.資料`（即「預設資料」）下的各項 `query` 方法實現為一個「通用查詢」：

- 一般的字/地位仍然直接查《廣韻》條目
- 一部分現代常用的字（尤其古今分合等差異較大者）作一定映射後再查出《廣韻》相應條目
- 此外引入若干必要（與廣韻字音差異大，或廣韻失收）的《切韻》《集韻》等常用字條目（切韻系限定，不在切韻系的暫緩引入）

這部分資料暫時不求多而全（但質量要夠），額外資料的格式也先自由一些（之後等到它有點規模或複雜程度後再看），先做出一版再說。

### 音韻地位

- 韻目「欣」改為「殷」（0.14.0 暫不允許寫「欣」，也許可以之後在小版本更新裡加回來（
- 元韻置於山攝（因為「攝」的時代元韻就是山攝，和《廣韻》韻目順序沒有關係（和《切韻》順序就更沒關係了（（

#### 2022-07-31 增加

- （重要）音韻地位編碼更新，不兼容之前版本（用 v0.13 編碼會得到錯誤結果
- 韻目「眞」改為更通用碼位「真」，與《廣韻》資料字頭、OpenCC 正字均一致（與「殷」類似，0.14.0 也暫不允許寫「眞」，之後再考慮
- 標準分析體系由 v2/v2ext 改名 std/ext，並作如下調整：
  - 虞：中立→合口
  - 侯：開口→中立
  - 幽：中立→開口

#### 2022-08-12 增加

音韻分析方面：

- 設立「前元音三等韻」分類：支脂祭真仙宵麻庚清幽侵鹽
- 設立「非前元音三等韻」分類：東鍾之微魚虞廢殷元文歌陽尤嚴凡
  - （臻蒸兩韻不定，以上兩類均不含此兩韻）
- 「重紐韻」增加「庚清蒸幽」四韻，即：支脂祭真仙宵庚清蒸幽侵鹽
  - 除蒸幽外，拼鈍音必須標重紐（只有蒸幽可標可不標，內置《廣韻》資料中部分小韻標重紐）
  - 庚三僅限標B類，清韻在 ext 體系允許B類，但正則化後（即 std 體系）會成為「庚三B」
  - 最簡描述中可以省略「庚三B」「清A」中的重紐
- 取消「重紐母」分類，其等同於「鈍音」
  - 云母亦為鈍音，拼重紐韻時亦需標重紐，且限B類（蒸幽韻亦同上述，可標可不標，內置《廣韻》資料「域」小韻不標）
  - 最簡描述中（蒸幽韻亦除外）可以省略云母的「B」
- 取消「輕脣韻」分類（`.屬於` 表達式中亦移除），因為並不如 `` is`脣音 三等 ${非前元音三等韻}` `` 覆蓋得更完整且邊界明確

標準分析體系與資料方面：

- 清韻、庚韻三等鈍音如上所述，現均要求標重紐
- 云母拼重紐韻時（蒸幽韻除外）如上所述，現要求標重紐
- 幽韻、蒸韻重紐相關：
  - 脣音/合口：v2(ext) 不允許「不標」與「B類」對立（標B會視同不標），現兩者均允許（內置《廣韻》資料中此類無標B者）
  - 其他鈍音：v2(ext) 不允許「不標」與「A類」對立（標A會視同不標），現兩者均允許（內置《廣韻》資料中此類無標A者）
- 麻韻三等鈍音亦允許標重紐，A/B 類不限，也可以不標（內置《廣韻》資料僅「乜」一小韻，標A類）

### 推導方案「兼容模式」

- v0.12 方案（不使用編碼查表的話）大體上仍可以在 v0.14 仍可（修改韻目用字後）以「兼容模式」運作，惟涉及「山臻攝」與「元韻」時可能需要調整一下
- v0.13 方案基本上可以繼續在 v0.14 運作，但需要仔細檢查與「虞」「侯」「幽」韻的「呼」相關的判斷條件並適當更新

### 「分析體系」相關

(2022-08-25 增加）

為了避免混淆，`音韻地位` 對象在 v0.14 將僅支持 `std` 與 `ext` 兩體系（分別取代之前的 `v2`、`v2ext`），不再支持表達其他體系音韻地位。

因此 `適配分析體系` 模塊也可以拆解為（或併入）`音韻地位.驗證`、`音韻地位.導入`、`音韻地位.正則化`。

別家體系仍可用含音韻地位六要素的普通對象（`object`）表示，不再享受 `音韻地位` 提供的各項功能【但可以用 `音韻地位.導入` 轉換為 ext 體系的 `音韻地位`

【而 `屬於` 也許也可以提供一個靜態函數的版本，能用於別家地位，但用於 `音韻地位` 以外的對象時，功能會相當受限，比如僅可直接判斷六要素，不可用諸如 `章組`、`鈍音`、`舒聲` 之類的高級判斷

【`適配分析體系` 原有的「轉換為其他家體系」功能也可以不用立即扔掉，未來修一修也許可以實現 `音韻地位.導入` 的逆操作

**Comment by graphemecluster** [comment]
今天剛想想了我這邊的計劃。我會先將推動器 0.2 的 Schema 部分搬到這兒（或是新 repo？），再把 `qieyun-autoderiver-evaluate` 搬回 `qieyun-autoderiver`，做好優化後才正式發佈。可我想我要到６月中旬至下旬才有空，所以你們也可以選擇幫我先改改，我不介意的，我也想快點正式發佈新版推導器。

回到這裏，我個人是覺得不需要拆開成兩個獨立 package 啦，即使真的要分開兩個 package 我也覺得 monorepo 比較好。

**Reply by syimyuzya** [reply]
> 我個人是覺得不需要拆開成兩個獨立 package 啦，即使真的要分開兩個 package 我也覺得 monorepo 比較好。

嘛～這個其實是因為我預期推導方案部分的 API 變動會比 qieyun 快很多（尤其是因為它的功能還有好多行為細節還沒討論過），分成兩個 repo 正是為了讓兩者能各自 release 😂

**Comment by untunt** [comment]
### 音韻地位

- 眞韻→真韻（如同原计划的 https://github.com/nk2028/qieyun-encoder-python/commit/81bcf30e366fb0428ea3caf49dd3db8bafbb2921 ）
- 開合改動
  - 虞：中立→合口（無需討論）
  - 侯：開口→中立（《毛詩音》反切以合口爲主）
  - 幽：中立→開口（《毛詩音》反切爲開口。可以認爲“中立僅限非前元音韻”優先於“尤侯幽要一致”）
    以下跟《毛詩音》反切不符，但保持原貌，認爲“-w、-m 尾韻算開口”更優先
  - 豪：開口（《毛詩音》反切開合各半）
  - 覃：開口（《毛詩音》反切爲合口）

**Reply by syimyuzya** [reply]
哇好耶～

**Reply by syimyuzya** [reply]
想了一下這樣的話，v0.14 會修改一下「編碼」格式，和 v0.13 會不兼容（但看起來問題不會很大w（而且新的編碼格式對「呼」和「重紐」的處理也更不依賴特定聲韻，一定程度上能用得更久些w

**Comment by untunt** [comment]
法（SyiMyuZya）餛（unt）討論：

- 音韻地位修改
  - 云母拼重紐韻時一律標 B 類，在最簡描述中省略
  - 清韻鈍音一律標 A 類，庚韻三等鈍音一律標 B 類，在最簡描述中省略
  - 乜小韻 *miaeq* 標 A 類
- [`聲韻搭配.ts`](https://github.com/nk2028/qieyun-js/blob/main/src/lib/%E8%81%B2%E9%9F%BB%E6%90%AD%E9%85%8D.ts) 修改
  - `重紐母` 移除
  - `重紐韻` 增加庚清蒸幽，即改爲：支脂祭眞<真>仙宵庚清蒸幽侵鹽
  - 新增 `前元音三等韻`：支脂祭眞<真>仙宵麻庚清幽侵鹽
  - 新增 `非前元音三等韻`：東鍾之微魚虞廢欣<殷>元文歌陽尤嚴凡
    （臻蒸二韻地位不定，所以在兩項中都不包含）

說明：
- 云母儘管沒有 A 類，但顯示 B 類是爲了滿足「鈍音聲母 ⇔ 重紐母」，以及防止初學者誤解爲 C 類。同時 `重紐母` 因和表達式中的 `鈍音` 重複而移除
- 庚<sub>三</sub>清鈍音顯示重紐類也是爲了防止初學者誤解爲 C 類
- 麻蒸幽韻默認不指定重紐類，因爲不確定，只有確定的一些小韻明確指定重紐類。其中麻韻鈍音僅一個乜小韻，它恰好也是可以明確指定重紐類的小韻
- 蒸幽韻在《切韻》中有重紐對立，麻韻無，所以蒸幽韻加入 `重紐韻` 而麻韻不加入
- `前元音三等韻` 和 `非前元音三等韻` 均不含臻蒸，因爲這兩個韻地位不確定
- 非前元音三等韻包含了「輕唇韻」且更有意義，所以無需再增加「輕唇韻」
- 這些常量均不需要加入表達式
